client/src/components/Sharing/Sharing.vue

```diff
@@ -0,0 +1,593 @@
+            return item.extra && (item.extra.can_change.length > 0 || item.extra.cannot_change.length > 0);
+        },
+        onMultiselectBlur(isAdmin) {
+            const isValueChosen = this.multiselectValues.sharingCandidates.some(
+                (item) => item.email === this.multiselectValues.currentUserSearch
+            );
+            if (this.multiselectValues.currentUserSearch && !isValueChosen && !isAdmin) {
+                this.multiselectValues.sharingCandidates.push({ email: this.multiselectValues.currentUserSearch });
+            }
+        },
+        addError(newError) {
+            // temporary turning Set into Array, until we update till Vue 3.0, that supports Set reactivity
+            this.errors = Array.from(new Set(this.errors).add(newError));
+        },
+        onCopy() {
+            copy(this.itemUrl);
+            this.tooltipClipboard = "Copied!";
+        },
+        onCopyOut() {
+            this.tooltipClipboard = "Copy URL";
+        },
+        onEdit() {
+            this.showUrl = false;
+        },
+        assignItem(newItem, overwriteCandidates) {
+            if (newItem.errors) {
+                this.errors = newItem.errors;
+            }
+            this.item = newItem;
+            if (overwriteCandidates) {
+                this.multiselectValues.sharingCandidates = Array.from(newItem.users_shared_with);
+            }
+            if (!this.item.extra || newItem.errors.length > 0) {
+                this.item.extra = defaultExtra();
+            }
+
+            this.ready = true;
+        },
+        onChange(newSlug) {
+            this.showUrl = true;
+            const requestUrl = `${this.slugUrl}`;
+            axios
+                .put(requestUrl, {
+                    new_slug: newSlug,
+                })
+                .then(() => (this.item.username_and_slug = `${this.itemSlugParts[0]}${newSlug}`))
+                .catch((error) => this.addError(error));
+        },
+        onImportable(importable) {
+            if (importable) {
+                this.setSharing(this.actions.enable_link_access);
+            } else {
+                this.item.published = false;
+                this.setSharing(this.actions.disable_link_access);
+            }
+        },
+        onPublish(published) {
+            if (published) {
+                this.item.importable = true;
+                this.setSharing(this.actions.publish);
+            } else {
+                this.setSharing(this.actions.unpublish);
+            }
+        },
+        getModel() {
+            this.ready = false;
+            this.dismissCountDown = 0;
+            axios
+                .get(`${getAppRoot()}api/${this.pluralNameLower}/${this.id}/sharing`)
+                .then((response) => this.assignItem(response.data, true))
+                .catch((error) => this.addError(error.response.data.err_msg));
+        },
+        setUsername() {
+            const Galaxy = getGalaxyInstance();
```

Comment:
I don't think `setUsername` should be the responsibility of the sharing component. If you move this out into its own component I think that would look better.